vlasovsolver/gpu_pitch_angle_diffusion.cpp

```diff
@@ -0,0 +1,799 @@
+   int blocksPerGrid = numberOfLocalCells;
+   int sharedMemorySize = totalThreadsPerBlock * sizeof(Realf);
+
+   if (getObjectWrapper().particleSpecies[popID].sparse_conserve_mass) {
+      // Ensure mass conservation
+      calculateDensity_kernel<<<blocksPerGrid, threadsPerBlock, sharedMemorySize>>>(
+         dev_densityPreAdjust,
+         dev_VBCs
+      );
+      
+      CHK_ERR( gpuPeekAtLastError() );
+      CHK_ERR( gpuDeviceSynchronize() );
+   }
+
+   std::vector<Real> dtTotalDiff(numberOfLocalCells, 0.0); // Diffusion time elapsed for each spatial cells
+
+   while (!allSpatialCellTimeLoopsComplete) { // Substep loop
+
+      // Compute maximum indices
+      int maxBlockIndex = 0;
+      int maxCellIndex = 0;
+      for (size_t CellIdx = 0; CellIdx < numberOfLocalCells; CellIdx++) { // Iterate over all spatial cells
+         if(spatialLoopComplete[CellIdx]){
+            continue;
+         }
+
+         const auto CellID                  = LocalCells[CellIdx];
+         SpatialCell& cell                  = *mpiGrid[CellID];
+
+         maxBlockIndex += cell.get_number_of_velocity_blocks(popID);
+         maxCellIndex++;
+      } // End spatial cell loop
+
+      int maxThreadsPerBlock = Hashinator::defaults::MAX_BLOCKSIZE;
+      int blocksPerVelocityCell = (nbins_v*nbins_mu+maxThreadsPerBlock-1)/maxThreadsPerBlock;
+
+      // Construct cellIdx arrays
+      std::vector<size_t> host_cellIdxStartCutoff(maxCellIndex);
+      std::vector<size_t> host_smallCellIdxArray(maxCellIndex);
+      std::vector<size_t> host_remappedCellIdxArray(numberOfLocalCells); // The position of the cell index in the sequence instead of the actual index
+
+      // Calculate indices
+      int remappedCellIdx = 0;
+      vmesh::LocalID totalNumberOfVelocityBlocks = 0;
+      for (size_t CellIdx = 0; CellIdx < numberOfLocalCells; CellIdx++) { // Iterate over all spatial cells
+         // Add elements to remapped cellIdx array
+         host_remappedCellIdxArray[CellIdx] = remappedCellIdx;
+
+         if(spatialLoopComplete[CellIdx]){
+            continue;
+         }
+
+         const auto CellID                  = LocalCells[CellIdx];
+         SpatialCell& cell                  = *mpiGrid[CellID];
+
+         vmesh::LocalID numberOfVelocityBlocks = cell.get_number_of_velocity_blocks(popID);
+
+         // Add elements to cellIdx arrays
+         host_cellIdxStartCutoff[remappedCellIdx] = totalNumberOfVelocityBlocks;
+         host_smallCellIdxArray[remappedCellIdx] = CellIdx;
+
+         totalNumberOfVelocityBlocks += numberOfVelocityBlocks;
+         remappedCellIdx++;
+      } // End spatial cell loop
+
+      std::vector<Real> host_Ddt (maxCellIndex);
+
+      // Create device arrays
+      size_t *dev_cellIdxArray, *dev_velocityIdxArray, *dev_smallCellIdxArray, *dev_remappedCellIdxArray, *dev_cellIdxStartCutoff;
+      Real *dev_Ddt, *dev_potentialDdtValues;
+      int *dev_cellIdxKeys;
+
+      // Allocate memory
```

Comment:
Ah, more transient cuda mallocs? These could also be persistent and included in overall GPU memory consolidation. In particular as this is inside a while loop!